mets_adapter/mets_adapter/src/main/scala/uk/ac/wellcome/mets_adapter/services/BagRetriever.scala

```diff
@@ -9,18 +9,23 @@ import akka.stream.ActorMaterializer
 import de.heikoseeberger.akkahttpcirce.FailFastCirceSupport._
 import grizzled.slf4j.Logging
 import io.circe.generic.auto._
-import uk.ac.wellcome.mets_adapter.models.Bag
+import uk.ac.wellcome.mets_adapter.models._
 
 import scala.concurrent.{ExecutionContext, Future}
 
-class BagRetriever(url: String, tokenService: TokenService)(
+trait BagRetriever {
```

Comment:
I think I'm missing something here, but what's the reason for having a trait and a class implementing it?

Reply:
I did this for 2 (related) reasons:

* For unit testing, it gets messy to always have to mock HTTP calls: so for example when testing the `MetsAdapterService` we don't care about the details of how the `BagRetriever` works, just that it is something that returns bags. By having it as a trait we easily [create a dummy bag retriever](https://github.com/wellcometrust/catalogue/pull/261/files#diff-f84d662997f5ad022e6813a899bdee93R41-R44) without having to resort to mocking frameworks etc

* Beyond just testing I think it's better to program against interfaces rather than specific implementations (I'm sure you understand this, but to state the obvious it makes it easier to swap things out in future, and also helps clearly separate any implementation details from the core functionality of a class)